query/worker.go

```diff
@@ -0,0 +1,239 @@
+// Results for any query handled by this worker will be delivered on the
+// results channel. quit can be closed to immediately make the worker exit.
+//
+// The method is blocking, and should be started in a goroutine. It will run
+// until the peer disconnects or the worker is told to quit.
+//
+// NOTE: Part of the Worker interface.
+func (w *worker) Run(peer Peer, results chan<- *jobResult,
+	quit <-chan struct{}) {
+
+	// Subscribe to messages from the peer.
+	msgChan, cancel := peer.SubscribeRecvMsg()
+	defer cancel()
+
+	for {
+		log.Tracef("Worker %v waiting for more work", peer.Addr())
+
+		var job *queryJob
+		select {
+		// Poll a new job from the nextJob channel.
+		case job = <-w.nextJob:
+			log.Tracef("Worker %v picked up job with index %v",
+				peer.Addr(), job.Index())
+
+		// Ignore any message received while not working on anything.
+		case msg := <-msgChan:
+			log.Tracef("Worker %v ignoring received msg %T "+
+				"since no job active", peer.Addr(), msg)
+			continue
+
+		// If the peer disconnected, we can exit immediately, as we
+		// weren't working on a query.
+		case <-peer.OnDisconnect():
+			log.Debugf("Peer %v for worker disconnected",
+				peer.Addr())
+			return
+
+		case <-quit:
+			return
+		}
+
+		select {
+		// There is no point in queueing the request if the job already
+		// is canceled, so we check this quickly.
+		case <-job.cancelChan:
+			log.Tracef("Worker %v found job with index %v "+
+				"already canceled", peer.Addr(), job.Index())
+
+		// We received a non-canceled query job, send it to the peer.
+		default:
+			log.Tracef("Worker %v queuing job %T with index %v",
+				peer.Addr(), job.Req, job.Index())
+
+			peer.QueueMessageWithEncoding(job.Req, nil, job.encoding)
+		}
+
+		// Wait for the correct response to be received from the peer,
+		// or an error happening.
+		var (
+			resp    wire.Message
+			jobErr  error
+			timeout = time.After(job.timeout)
+		)
+
+	Loop:
+		for {
+			select {
+			// A message was received from the peer, use the
+			// response handler to check whether it was answering
+			// our request.
+			case resp = <-msgChan:
+				progress := job.HandleResp(
+					job.Req, resp, peer.Addr(),
+				)
+
+				log.Tracef("Worker %v handled msg %T while "+
+					"waiting for response to %T (job=%v). "+
+					"Finished=%v, progressed=%v",
+					peer.Addr(), resp, job.Req, job.Index(),
+					progress.Finished, progress.Progressed)
+
+				// If the response did not answer our query, we
+				// check whether it did progress it.
+				if !progress.Finished {
+					// If it did make progress we reset the
+					// timeout. This ensures that the
+					// queries with multiple responses
+					// expected won't timeout before all
+					// responses have been handled.
+					// TODO(halseth): separete progress
+					// timeout value.
+					if progress.Progressed {
+						timeout = time.After(
+							job.timeout,
+						)
+					}
+					continue Loop
```

Comment:
What happens to the response in this case if we continue? I guess it's OK since it's been handed to the caller already via the `HandleResp` method?

Reply:
Yeah, it will have been handled by the response handler, and we only need to reset the timeout to ensure progressing queries don't timeout.

Reply:
Looks like we could limit the scope of `resp` to only be within the `select` statement then? Its current position makes it seem as if it will be used later on.

Reply:
Yes, very good catch! Probably a leftover from when it was used as part of the response. Fixed! https://github.com/lightninglabs/neutrino/pull/179/commits/20edcf05830ee6d4665f41bd98374df79eb8fb7a